backend/functions/index.js

```diff
@@ -52,3 +53,29 @@ exports.updateMostRecentMessage = functions.database.ref('/messages/{messageId}'
 
 		return ref.child(message.conversation).child("lastMessage").set(message);
 	});
+
+/**
+ * Sets a unique username for the newly created user
+ */
+exports.setUniqueUsername = functions.auth.user().onCreate(event => {
+	var ref = admin.database().ref("users");
+
+	const id = event.data.uid;
+	const email = event.data.email;
+	// string before '@' symbol with trailing digits removed
+	const potentialUsername = email.substring(0, email.indexOf('@')).replace(/\d+$/, '');
+
+	ref.once("value", function(snapshot) {
+		var matchedUsernames = 0;
+		snapshot.forEach(function(data) {
+			const user = data.val();
+			if (user.username !== undefined && user.username.startsWith(potentialUsername)) {
```

Comment:
I think this will match items that aren't part of your name-deduplication scheme.  What if the database already has a user named 'bob1' (created with email `bob1@yahoo.com`) and a new user with email `bob@gmail.com` tries to subscribe?

Probably better to generate potential names and test whether they are already in use.